python/mlc_llm/bench/executor.py

```diff
@@ -147,9 +153,147 @@ def get_executor_feature_dict(self) -> Dict[str, Any]:
+        if len(request_records) == 0:
+            return []
+
+        async def process_task_impl(
+            f_create_api_endpoint: Callable[[], APIEndPoint],
+            request_records: List[RequestRecord],
+            base_timestamp: float,
+            base_sys_time: float,
+            launch_start_gap_time: float,
+        ) -> List[RequestRecord]:
+            api_endpoint = f_create_api_endpoint()
+            loop = asyncio.get_running_loop()
+            # Get the delta time to convert system time to the loop time.
+            # We must use the system time `time.time()` which is consistent across processes.
+            loop_sys_delta_time = loop.time() - time.time()
+            updated_request_records: List[RequestRecord] = []
+            async with api_endpoint:
+
+                async def _task(request_record: RequestRecord) -> None:
+                    updated_request_records.append(await api_endpoint(request_record))
+
+                tasks = []
+                for request_record in request_records:
+                    launch_time = (
+                        (request_record.timestamp - base_timestamp)
+                        + (base_sys_time + launch_start_gap_time)
+                        + loop_sys_delta_time
+                    )
+                    loop.call_at(
+                        launch_time,
+                        lambda record: tasks.append(asyncio.create_task(_task(record))),
+                        request_record,
+                    )
+                    # Sleep to allow runs of other scheduled tasks if any.
+                    await asyncio.sleep(0)
+
+                # Sleep until all the tasks are launched.
+                await asyncio.sleep(launch_time - loop.time() + launch_start_gap_time)
+                # Wait for all tasks to be scheduled
+                assert len(tasks) == len(request_records)
+                await asyncio.gather(*tasks)
+
+            assert len(updated_request_records) == len(request_records)
+            return updated_request_records
+
+        return asyncio.run(
+            process_task_impl(
+                f_create_api_endpoint,
+                request_records,
+                base_timestamp,
+                base_sys_time,
+                launch_start_gap_time,
+            )
+        )
+
+    async def warmup(self, warmup_requests: List[RequestRecord]) -> None:
+        # Disable tqdm for warmup
+        disable_tqdm = self.disable_tqdm
+        self.disable_tqdm = True
+        # Send all warmup requests together by resetting timestamps to 0.
+        for request_record in warmup_requests:
+            request_record.timestamp = 0.0
+        await self.run_benchmark(warmup_requests)
+        self.disable_tqdm = disable_tqdm
+
+    def get_num_warmup_requests(self) -> int:
+        # Use 32 requests for warmup at most.
+        return min(self.num_requests, 32)
+
+    def get_request_rate(self) -> Optional[np.float32]:
```

Comment:
remove request_rate, since the request_rate is only a function of how to generate List of record, executor should be independent from this